plat/rpi/rpi5/scmi/scmi_reset.c

```diff
@@ -7,31 +7,143 @@
 #include <stdint.h>
 
 #include <common/debug.h>
+#include <drivers/delay_timer.h>
 #include <drivers/scmi.h>
 #include <lib/utils_def.h>
 #include <platform_def.h>
+#include <lib/mmio.h>
+
+#include "rpi5_scmi_resources.h"
+
+#define SW_INIT_SET		0x00
+#define SW_INIT_CLEAR		0x04
+#define SW_INIT_STATUS		0x08
+
+#define SW_INIT_BIT(id)		BIT((id) & 0x1f)
+#define SW_INIT_BANK(id)	((id) >> 5)
+
+#define SW_INIT_BASE U(0x1001504318)
+/* A full bank contains extra registers that we are not utilizing but still
+ * qualify as a single bank.
+ */
+#define SW_INIT_BANK_SIZE	0x18
+
+struct scmi_reset rpi5_resets[RPI5_SCMIRST_MAX] = {
+	[RPI5_SCMIRST_PCIE1_1] = {
+		.name = "swinit_pcie1",
+		.id = 7,
+	},
+	[RPI5_SCMIRST_PCIE1_2] = {
+		.name = "bridge_pcie1",
+		.id = 43,
+	},
+	[RPI5_SCMIRST_PCIE2_1] = {
+		.name = "swinit_pcie2",
+		.id = 32,
+	},
+	[RPI5_SCMIRST_PCIE2_2] = {
+		.name = "bridge_pcie2",
+		.id = 44,
+	},
+};
+
+static int brcmstb_reset_assert(unsigned int scmi_id)
+{
+	unsigned int off = SW_INIT_BANK(rpi5_resets[scmi_id].id) * SW_INIT_BANK_SIZE;
+
+	mmio_write_32(SW_INIT_BASE + off + SW_INIT_SET,
+			SW_INIT_BIT(rpi5_resets[scmi_id].id));
+
+	return 0;
+}
+
+static int brcmstb_reset_deassert(unsigned int scmi_id)
+{
+	unsigned int off = SW_INIT_BANK(rpi5_resets[scmi_id].id) * SW_INIT_BANK_SIZE;
+
+	mmio_write_32(SW_INIT_BASE + off + SW_INIT_CLEAR,
+			SW_INIT_BIT(rpi5_resets[scmi_id].id));
+
+	/* Maximum reset delay after de-asserting a line and seeing block
+	 * operation is typically 14us for the worst case, build some slack
+	 * here.
+	 */
+	udelay(100);
```

Comment:
Taking into account that this is ARM TF and you are blocking calling CPU in Secure state, maybe 14us will be sufficient?

Reply:
I think it probably should not be here at all, or should per-reset cfg in rpi5_resets